core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/TripleStore.java

```diff
@@ -859,41 +863,59 @@ public void startTransaction() throws IOException {
 	 */
 	void endTransaction(boolean commit) throws IOException {
 		if (writeTxn != 0) {
-			if (commit) {
-				E(mdb_txn_commit(writeTxn));
-				if (recordCache != null) {
-					StampedLock lock = txnManager.lock();
-					long stamp = lock.writeLock();
+			try {
+				if (commit) {
 					try {
-						txnManager.deactivate();
-						mapSize = LmdbUtil.autoGrowMapSize(mapSize, pageSize, 0);
-						E(mdb_env_set_mapsize(env, mapSize));
-						// restart write transaction
-						try (MemoryStack stack = stackPush()) {
-							PointerBuffer pp = stack.mallocPointer(1);
-							mdb_txn_begin(env, NULL, 0, pp);
-							writeTxn = pp.get(0);
-						}
-						updateFromCache();
-						// finally, commit write transaction
 						E(mdb_txn_commit(writeTxn));
-					} finally {
-						recordCache = null;
-						try {
-							txnManager.activate();
-						} finally {
-							lock.unlockWrite(stamp);
+						if (recordCache != null) {
+							StampedLock lock = txnManager.lock();
```

Comment:
I'm just wondering if any of the variables you are checking prior to this lock need to be volatile.